### ISSUE TITLE
this image loop print too long logs

### ISSUE THREAD

**Issue by hetykai** [0rzech/gitea-docker#1]
`docker logs gitea`
Loop print log:
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@         WARNING: UNPROTECTED PRIVATE KEY FILE!          @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
Permissions 0755 for '/data/ssh/ssh_host_ed25519_key' are too open.
It is required that your private key files are NOT accessible by others.
This private key will be ignored.
key_load_private: bad permissions
Could not load host key: /data/ssh/ssh_host_ed25519_key
sshd: no hostkeys available -- exiting.

Like this:
![powershell_2018-05-13_10-50-37](https://user-images.githubusercontent.com/3378432/39963373-04102134-569c-11e8-9aa3-6005759b2793.png)

And i search "ssh_host_ed25519_key" in gitea's Official repository, i found two files is different with this repository, is that the problem?

**Comment by 0rzech** [comment]
I guess you have discovered the root cause yourself, but I will reply just in case.

SSH daemon (sshd) expects keys to be readable only by owner and nobody else. That's why ssh init script creates ssh directory with 0700 and key with 0600 permissions (see [here](https://github.com/0rzech/gitea-docker/blob/bf5b1c395abfe2b6d7b6410c0b44306e3ae6d3c7/files/etc/s6/openssh/run)).

You must have had either created keys yourself with wrong permissions or accidentally changed those and sshd started complaining about it.

**Comment by 0rzech** [comment]
I guess you have discovered the root cause yourself, but I will reply just in case.

SSH daemon (sshd) expects keys to be readable only by owner and nobody else. That's why ssh init script creates ssh directory with 0700 and key with 0600 permissions (see [here](https://github.com/0rzech/gitea-docker/blob/bf5b1c395abfe2b6d7b6410c0b44306e3ae6d3c7/files/etc/s6/openssh/run)).

You have probably either created keys yourself with wrong permissions or accidentally changed those and sshd started complaining about it.

**Comment by 0rzech** [comment]
I guess you have discovered the root cause yourself, but I will reply just in case.

SSH daemon (sshd) expects keys to be readable only by their owner and nobody else. That's why ssh init script creates ssh directory with 0700 and key with 0600 permissions (see [here](https://github.com/0rzech/gitea-docker/blob/bf5b1c395abfe2b6d7b6410c0b44306e3ae6d3c7/files/etc/s6/openssh/run)).

You have probably either created keys with wrong permissions yourself or accidentally changed permissions afterwards and sshd started complaining about it.

> And i search "ssh_host_ed25519_key" in gitea's Official repository, i found two files is different with this repository, is that the problem?

There's no problem with that. Official Gitea container uses more key types (some of which are less secure), so that clients with older versions of SSH can still establish connection. I use Ed25519 algorithm only, which the newest one. Also, official container provides key size when generating Ed25519 key, which, according to ArchWiki, is unnecessary:
> There is no need to set the key size, as all Ed25519 keys are 256 bits. Also, they rely on a new key format which "uses a bcrypt-based key derivation function that makes brute-force attacks against stolen private keys far slower". [[source](https://wiki.archlinux.org/index.php/SSH_keys#Ed25519)]

**Comment by 0rzech** [comment]
> (...) Permissions 0755 for '/data/ssh/ssh_host_ed25519_key' are too open. (...)

I guess you have discovered the root cause yourself, but I will reply just in case.

SSH daemon (sshd) expects keys to be readable only by their owner and nobody else. That's why ssh init script creates ssh directory with 0700 and key with 0600 permissions [1].

You have probably either created keys with wrong permissions yourself or accidentally changed permissions afterwards and sshd started complaining about it.

> And i search "ssh_host_ed25519_key" in gitea's Official repository, i found two files is different with this repository, is that the problem?

Official Gitea container uses more key types, some of which are less secure [2]:
> Ed25519 is an elliptic curve signature scheme that offers better security than ECDSA and DSA and good performance.

In return, clients with older versions of SSH can still establish a connection. I use Ed25519 algorithm only. Also, official container provides key size when generating Ed25519 key, which is unnecessary [2]:
> There is no need to set the key size, as all Ed25519 keys are 256 bits. Also, they rely on a new key format which "uses a bcrypt-based key derivation function that makes brute-force attacks against stolen private keys far slower". 

In short: I don't see any problem with that.

---
Sources:
[1] [ gitea-docker/files/etc/s6/openssh/run](/gitea-docker/blob/bf5b1c395abfe2b6d7b6410c0b44306e3ae6d3c7/files/etc/s6/openssh/run)
[2] [SSH keys - ArchWiki](https://wiki.archlinux.org/index.php/SSH_keys#Ed25519)

**Comment by 0rzech** [comment]
> (...) Permissions 0755 for '/data/ssh/ssh_host_ed25519_key' are too open. (...)

I guess you have discovered the root cause yourself, but I will reply just in case.

SSH daemon (sshd) expects keys to be readable only by their owner and nobody else. That's why ssh init script creates ssh directory with 0700 and key with 0600 permissions [1].

You have probably either created keys with wrong permissions yourself or accidentally changed permissions afterwards and sshd started complaining about it.

> And i search "ssh_host_ed25519_key" in gitea's Official repository, i found two files is different with this repository, is that the problem?

Official Gitea container uses more key types, some of which are less secure [2]:
> Ed25519 is an elliptic curve signature scheme that offers better security than ECDSA and DSA and good performance.

In return, clients with older versions of SSH can still establish a connection. I use Ed25519 algorithm only. Also, official container provides key size when generating Ed25519 key, which is unnecessary [2]:
> There is no need to set the key size, as all Ed25519 keys are 256 bits. Also, they rely on a new key format which "uses a bcrypt-based key derivation function that makes brute-force attacks against stolen private keys far slower". 

In short: I don't see any problem with that.

---
Sources:
[1] [gitea-docker/files/etc/s6/openssh/run](/0rzech/gitea-docker/blob/bf5b1c395abfe2b6d7b6410c0b44306e3ae6d3c7/files/etc/s6/openssh/run)
[2] [SSH keys - ArchWiki](https://wiki.archlinux.org/index.php/SSH_keys#Ed25519)

**Comment by hetykai** [comment]
I found the cause of this issue because of the Docker volume permissions issue in Windows. Do not mount /data because it contains /data/ssh/ssh_host_ed25519_key. That is, create container with like `-v d:/gitea/git:/data/git -v d:/gitea/gitea:/data/gitea`. This is not this image‘s issue, so I closed it. Still thank you for your reply.

**Comment by 0rzech** [comment]
You're welcome!
